lib/dry/equalizer.rb

```diff
@@ -74,8 +75,16 @@ def define_cmp_method
     # @api private
     def define_hash_method
       keys = @keys
-      define_method(:hash) do | |
-        keys.map(&method(:send)).push(self.class).hash
+      immutable = @immutable
+
+      if immutable
+        define_method(:hash) do | |
+          @hash ||= keys.map(&method(:send)).push(self.class).hash
+        end
+      else
+        define_method(:hash) do
+          keys.map(&method(:send)).push(self.class).hash
```

Comment:
I believe I considered doing that back then but I didn't like the idea of having the `if` running everytime `hash` was called. Probably not a big deal, but feels a little wrong to me. Another option would be to define a lambda with the hash logic and have both branches call it, wdyt?